File: ios-app/BayitPlusApp/Models/MovieInteraction.swift
import Foundation

struct InteractableMovieItem: Codable, Identifiable, Equatable, Hashable {
    let contentId: String
    let title: String
    let posterUrl: String?
    let characterCount: Int
    let interactionCount: Int
    let maxInteractions: Int
    let status: String

    var id: String {
        contentId
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        contentId = try container.decode(String.self, forKey: .contentId)
        title = try container.decode(String.self, forKey: .title)
        posterUrl = try container.decodeIfPresent(String.self, forKey: .posterUrl)
        characterCount = try container.decode(Int.self, forKey: .characterCount)
        interactionCount = try container.decodeIfPresent(Int.self, forKey: .interactionCount) ?? 0
        maxInteractions = try container.decodeIfPresent(Int.self, forKey: .maxInteractions) ?? 10
        status = try container.decode(String.self, forKey: .status)
    }
}

struct MovieTagStatusItem: Codable {
    let contentId: String
    let status: String
    let characters: [InteractiveCharacterItem]
    let error: String?
}

struct InteractiveCharacterItem: Codable, Identifiable, Equatable, Hashable {
    let name: String
    let voiceId: String
    let frameUrl: String
    let description: String
    let movieContext: String
    let actorName: String?
    let gender: String?
    let suggestedQuestions: [String]

    var id: String {
        name
    }
}

struct CharacterQuestionsItem: Codable {
    let characterName: String
    let specificQuestions: [String]
    let genericQuestions: [String]
}

struct InteractionSelectionRequest: Codable {
    let contentId: String
    let characterName: String
    let questions: [String]
}

struct InteractionSelectionResponse: Codable {
    let contentId: String
    let createdCount: Int
    let totalInteractionCount: Int
    let maxInteractions: Int
    let generationStatus: String
}

struct InteractionMomentStatusItem: Codable, Identifiable {
    let characterName: String
    let interactionPrompt: String
    let status: String
    let videoUrl: String?

    var id: String {
        "\(characterName)-\(interactionPrompt)"
    }
}

struct InteractionStatusResponse: Codable {
    let contentId: String
    let interactionCount: Int
    let maxInteractions: Int
    let moments: [InteractionMomentStatusItem]
}

File: backend/app/models/movie_interaction.py
"""
Movie Interaction Models

Request/response schemas for the movie interaction hub feature.
Users can tag movies for character extraction, browse characters,
and ask predefined or custom questions.
"""

from typing import List, Optional
from pydantic import BaseModel, Field
from app.models.vod_interaction import ContentCharacter


class MovieTagRequest(BaseModel):
    """Request to tag a movie for character extraction."""
    content_id: str = Field(..., description="Content document ID")
    profile_id: str = Field(..., description="Child profile ID")


class MovieTagStatus(BaseModel):
    """Status of a movie tagging/character extraction job."""
    content_id: str
    status: str = Field(..., description="pending, processing, ready, failed")
    characters: List[ContentCharacter] = Field(default_factory=list)
    error: Optional[str] = None


class CharacterQuestionsResponse(BaseModel):
    """Questions available for a specific character."""
    character_name: str
    specific_questions: List[str] = Field(
        default_factory=list,
        description="AI-generated character-specific questions",
    )
    generic_questions: List[str] = Field(
        default_factory=list,
        description="Template-based generic questions",
    )


class InteractableMovie(BaseModel):
    """A movie that has been tagged for character interactions."""
    content_id: str
    title: str
    poster_url: Optional[str] = None
    character_count: int = 0
    interaction_count: int = Field(
        default=0,
        description="Current number of interactive moments on this content",
    )
    max_interactions: int = Field(
        default=10,
        description="Max interactive moments allowed per content",
    )
    status: str = Field(
        default="ready",
        description="ready or processing",
    )


class InteractionSelectionRequest(BaseModel):
    """Request to select questions for video generation."""
    content_id: str = Field(..., description="Content document ID")
    character_name: str = Field(..., description="Character to generate for")
    questions: List[str] = Field(
        ...,
        min_length=1,
        description="Selected question texts to generate interactions for",
    )


class InteractionMomentStatus(BaseModel):
    """Status of a single interactive moment generation."""
    character_name: str
    interaction_prompt: str
    status: str = Field(
        ...,
        description="queued, generating_audio, generating_video, complete, failed",
    )
    video_url: Optional[str] = None


class InteractionSelectionResponse(BaseModel):
    """Response after selecting questions for generation."""
    content_id: str
    created_count: int
    total_interaction_count: int
    max_interactions: int
    generation_status: str = Field(
        ...,
        description="queued, processing, complete",
    )


class InteractionStatusResponse(BaseModel):
    """Full interaction status for a content item."""
    content_id: str
    interaction_count: int
    max_interactions: int
    moments: List[InteractionMomentStatus] = Field(default_factory=list)
